core/node/rpc/sync/client/syncer_set.go

```diff
@@ -211,473 +144,261 @@ func (ss *SyncerSet) unlockStream(streamID StreamId) {
+					req.BackfillingFailureHandler(st)
+				}
+			}(cookie, backfill.GetSyncId())
 		}
 	}
 
-	// 2. The second try for adding and backfilling streams with the provided node address in cookie.
-	{
-		addingFailures := make([]*SyncStreamOpStatus, 0, len(req.ToAdd))
-		addingFailuresHandler := func(status *SyncStreamOpStatus) {
-			if status.GetCode() != int32(Err_NOT_FOUND) && status.GetCode() != int32(Err_INTERNAL) {
-				req.AddingFailureHandler(status)
-				return
+	// Process removing streams
+	wg.Add(len(req.ToRemove))
+	for _, streamID := range req.ToRemove {
+		go func(streamID StreamId) {
+			defer wg.Done()
+			if st := ss.processRemovingStream(ctx, streamID); st != nil {
+				req.RemovingFailureHandler(st)
 			}
+		}(StreamId(streamID))
+	}
 
-			addingFailuresLock.Lock()
-			addingFailures = append(addingFailures, status)
-			addingFailuresLock.Unlock()
-		}
+	wg.Wait()
 
-		backfillingFailures := make([]StreamId, 0, len(req.ToBackfill))
-		backfillingFailuresHandler := func(status *SyncStreamOpStatus) {
-			if status.GetCode() != int32(Err_NOT_FOUND) && status.GetCode() != int32(Err_INTERNAL) {
-				req.BackfillingFailureHandler(status)
-				return
-			}
+	return nil
+}
 
-			backfillingFailuresLock.Lock()
-			backfillingFailures = append(backfillingFailures, StreamId(status.GetStreamId()))
-			backfillingFailuresLock.Unlock()
-		}
+func (ss *SyncerSet) processAddingStream(
+	ctx context.Context,
+	syncID string,
+	cookie *SyncCookie,
+	changeNode bool,
+) *SyncStreamOpStatus {
+	streamID := StreamId(cookie.GetStreamId())
 
-		removingFailures := make([][]byte, 0, len(req.ToRemove))
-		removingFailuresHandler := func(status *SyncStreamOpStatus) {
-			if status.GetCode() != int32(Err_NOT_FOUND) && status.GetCode() != int32(Err_INTERNAL) {
-				req.RemovingFailureHandler(status)
-				return
-			}
+	if ss.otelTracer != nil {
+		var span trace.Span
+		ctx, span = ss.otelTracer.Start(ctx, "syncerset::processAddingStream",
+			trace.WithAttributes(attribute.String("streamId", streamID.String())),
+			trace.WithAttributes(attribute.String("syncId", syncID)))
+		defer span.End()
+	}
 
-			removingFailuresLock.Lock()
-			removingFailures = append(removingFailures, status.GetStreamId())
-			removingFailuresLock.Unlock()
-		}
+	unlock := ss.lockStream(streamID)
 
-		if err := ss.modify(ctx, ModifyRequest{
-			SyncID:                    req.SyncID,
-			ToBackfill:                req.ToBackfill,
-			ToAdd:                     req.ToAdd,
-			ToRemove:                  req.ToRemove,
-			BackfillingFailureHandler: backfillingFailuresHandler,
-			AddingFailureHandler:      addingFailuresHandler,
-			RemovingFailureHandler:    removingFailuresHandler,
-		}, false); err != nil {
-			return AsRiverError(err).Func("SyncerSet.Modify.second")
-		}
+	if _, found := ss.streamID2Syncer.Load(streamID); found {
+		unlock()
+		return ss.processBackfillingStream(ctx, syncID, syncID, cookie)
```

Comment:
After the unlock it is assumed that this stream is already part of the sync session. Is it possible that after unlock is called a remove stream requests comes in that drops the stream from syncing before the backfill request is handled?